### PR TITLE
Revamped Grabbing of Existentials

### PR DESCRIPTION
**Pull request by pedrotst**
This pull request makes a major rework on how existential grabbing works.

The main idea is to add a return statement to make sure that the grabbed existential will be properly cast into the return type of the underlying environment.

For example:

```
type 'a term =
  | T_Pair : 'a term * 'b term -> ('a * 'b) term
[@@coq_tag_gadt]

let rec interp : type a. a term -> a = function
  | T_Pair (p1, p2) -> (interp p1, interp p2)
```

Should be translated to:

```
Inductive term : vtag -> Set :=
| T_Pair : forall {a b : vtag}, term a -> term b -> term (tuple_tag a b)

Fixpoint interp {a : vtag} (function_parameter : term a) : decode_vtag a :=
  match function_parameter with
  | T_Pair p1 p2 =>
    let 'existT _ [__1, __0] [p2, p1] as exi :=
      existT (A := [vtag ** vtag]) (fun '[__1, __0] => [term __1 ** term __0])
        [_, _] [p2, p1]
      return
        let fst := projT1 exi in
        let __0 := Primitive.snd fst in
        let __1 := Primitive.fst fst in
        decode_vtag __0 * decode_vtag __1 in (* <--- We need to cast the return type to mention the grabbed existentials *)
    ((interp p1), (interp p2))
```

I understand that this pull request adds quite some verbosity to the code, but unfortunately it is necessary. I have plans to make this more robust in the future, having a full-fledged algorithm to only generate these existential when necessary. For now I'm implementing what is necessary for gadts as tags to work.

### CODE REVIEW

**Comment by clarus** [comment]
As a first comment, for your example:
```coq
Fixpoint interp {a : vtag} (function_parameter : term a) : decode_vtag a :=
  match function_parameter with
  | T_Pair p1 p2 =>
    let 'existT _ [__1, __0] [p2, p1] as exi :=
      existT (A := [vtag ** vtag]) (fun '[__1, __0] => [term __1 ** term __0])
        [_, _] [p2, p1]
      return
        let fst := projT1 exi in
        let __0 := Primitive.snd fst in
        let __1 := Primitive.fst fst in
        decode_vtag __0 * decode_vtag __1 in (* <--- We need to cast the return type to mention the grabbed existentials *)
    ((interp p1), (interp p2))
```
I think it should be possible to use something like that in the `return` (not tested):
```coq
 return
        let 'existT _ [__1, __0] _ := exi in
        ...
```

File: src/type.ml [hunk 1]
```diff
@@ -674,10 +821,105 @@ and tag_typ_constr
...
+            Name.Set.diff new_typ_args (Name.Set.of_list typ_args)
+          ))
+      )
+      []
+      (List.map typ_args_of_typ record_params) in
+  let new_typ_vars = VarEnv.reorg typ_args new_typ_vars in
+  return (record_params, new_typ_vars)
+
+
+and typed_existential_typs_of_typ
```
Comment (clarus) [inline src/type.ml:849]: We a comment for what does this function do (I should comment more my code too).

File: src/exp.ml [hunk 2]
```diff
@@ -819,12 +864,13 @@ and of_match :
...
     ) in
   let t = Match (e, dep_match, cases, is_with_default_case) in
+  (* If its a deppendent pattern matching then add eq_refl at the end of the match *)
   match dep_match with
   | None -> return t
   | Some dep_match ->
     let eq_refl = "eq_refl" |> Name.of_string_raw |> MixedPath.of_name in
     let ts = List.map (fun _ -> Some (Variable (eq_refl, []))) dep_match.args in
-  return (Apply (t, ts))
+    return (Apply (t, ts))
```
[the remark this answers is not in the document]
Comment (pedrotst) [reply src/exp.ml]: For this I would need to close this pull request, perform a rebase to delete this one commit and then re-open the pull request. I can do it, but its a lot of work for cleaning one single commit. Do you really need that?

On a side note I'll be more careful next time

[the remark this answers is not in the document]
Comment (clarus) [reply src/exp.ml]: Ah! You should not need to reopen the pull request to remove a commit, if you do a `git push -f` after the rebase.

**Pull request merged**